### DISCUSSION TITLE
J2000 Coordinate Frame

### DISCUSSION THREAD

**Discussion by jonsberndt** [JSBSim-Team/jsbsim#533] [Ideas]
JSBSim can be used to do some fun and educational orbital modeling and simulation. One of those interesting things is to model the International Space Station (ISS) in orbit. The current and forecasted ISS position in space can be found at this web site:

https://spotthestation.nasa.gov/trajectory_data.cfm

Data is given in J2000 frame - an inertial frame. I've wondered if we can convert from J2000 frame to our ECEF frame. Seems to me that the difference in the frames is essentially just an Earth rotation angle about the Z axis. 

The SPICE toolkit might be useful in implementing this. (See: https://naif.jpl.nasa.gov/naif/toolkit_C_PC_Cygwin_GCC_64bit.html)

Is this of interest to anyone else - has anyone else given this some thought?

**Comment by seanmcleod** [comment]
> Seems to me that the difference in the frames is essentially just an Earth rotation angle about the Z axis.

Guess it depends on the level of accuracy you want. So on Jan 1 2000 the Z-axis of the J2000 frame and the ECEF frames were aligned, but there has been precession and nutation since then.

See https://space.stackexchange.com/questions/51083/converting-j2000-to-ecef

I would be interested in seeing how accurately JSBSim could model the orbit of the ISS.

**Reply by bcoconni** [reply]
> I would be interested in seeing how accurately JSBSim could model the orbit of the ISS.

My uneducated guess (I'm no specialist in that area) is that JSBSim should be fine for a short term simulation but not so good for long term (JSBSim is missing models for extremely high altitude atmospheric properties (needed for drag), Moon gravity interaction, detailed Earth gravity such as [EGM96 or EGM2008 geoids](https://en.wikipedia.org/wiki/Earth_Gravitational_Model) and possibly some others).

Note that [GeographicLib provides a number of Earth gravity models](https://geographiclib.sourceforge.io/html/gravity.html) so the later could be easily added, keeping in mind that that would add some attachments that are quite large in size - 73Mb for the EGM2008 model !!!)

**Reply by bcoconni** [reply]
But I'd be interested as well :wink:

**Comment by seanmcleod** [comment]
> that JSBSim should be fine for a short term simulation..

Talking of short term versus longer term I noticed the following comment on the NASA link above:

> ISS Trajectory Operations and Planning Officer (TOPO)..... This thin atmosphere creates drag and over time can cause TOPO’s predicted ISS trajectory to accumulate error. Because of this, TOPO updates the predicted trajectory approximately three times a week,

So they update their predicted trajectory every 2-3 days, which seems pretty short term 😉 

But yes it would be interesting to see how quickly JSBSim's model diverges.

**Reply by bcoconni** [reply]
> So they update their predicted trajectory every 2-3 days, which seems pretty short term :wink:

Well, one can never be wrong with the expressions "short term" and "long term": I could pretend that my short term was shorter than you thought :smile: But indeed I would not have anticipated that atmospheric drag would affect simulations on such a short time span.

**Comment by dershow** [comment]
There are bunch of things that go into those models.  For example that thin atmosphere can vary a bunch with solar radiation.  When the sun is active the atmosphere grows. The Station control system also interacts with the drag.  For most operations The Station flies in LVLH (local vertical local horizontal).  So, the control system has to actively keep it pointed that way.  At other times, it can go into different modes.  And, even in LVLH, it’s possible to change how tightly it keeps to that attitude.  On the station control system it’s possible to tell if the astronauts are awake or not, because the control system responds to their movements.  Schedule an EVA and change control modes.  Detect some orbital debris and change the orbit.  And, the solar arrays track the sun, and have a relatively large affect on drag depending on their relative attitude to the velocity vector.  
So, a big part of the TOPOs is not about a passive object in orbit, but how it is controlled and how that makes it interact with the orbit.


--Adam

**Comment by seanmcleod** [comment]
Okay, well let's start of with just 4 minutes then 😉 

So the ephemeris gives 2 observation dates, 4 min apart, with `(x,y,z)` and `(vx,vy,vz)` in the J2000 frame.

```sh
2021-12-08T15:45:00.000 6555.630055870490 1256.781597809450 1289.667697856690 0.19784546610008 4.97053172606556 -5.82499523600153

2021-12-08T15:49:00.000 6364.220935726690 2389.526640678390 -138.318378509789 -1.78312287197086 4.41141843919986 -6.00205609106771
```

So if we use Spice or Astropy to convert both observations to the ECEF frame, then we can input the first observation as the JSBSim initial conditions and then after 4 mins of simulation compare the current ECEF state in JSBSim with the 2nd observation and see what the delta is in terms of position and velocity.

The NASA link also provides the required ISS data in terms of mass and drag, so we can update JSBSim's ball aircraft/create a new ISS aircraft with the data.

```sh
COMMENT Source: This file was produced by the TOPO office within FOD at JSC.
COMMENT Units are in kg and m^2
COMMENT MASS=454862.00
COMMENT DRAG_AREA=1492.30
COMMENT DRAG_COEFF=1.00
```

**Comment by seanmcleod** [comment]
The text file version of the ephemeris data doesn't explicitly list the units for position and velocity, but the XML version does.

```xml
<stateVector>
<EPOCH>2021-342T15:45:00.000Z</EPOCH>
<X units="km">6555.6300558704897</X>
<Y units="km">1256.78159780945</Y>
<Z units="km">1289.66769785669</Z>
<X_DOT units="km/s">0.19784546610008</X_DOT>
<Y_DOT units="km/s">4.9705317260655599</Y_DOT>
<Z_DOT units="km/s">-5.8249952360015298</Z_DOT>
</stateVector>
```
